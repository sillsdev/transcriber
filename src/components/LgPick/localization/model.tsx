// WARNING: This file is generated using ToModel.xsl. Changes made here may be lost.
import * as Localize from 'react-localization';

export interface ILanguagePickerStrings extends Localize.LocalizedStringsMethods {
    "font": string;
    "script": string;
    "language": string;
    "selectLanguage": string;
    "findALanguage": string;
    "codeExplained": string;
    "subtags": string;
    "details": string;
    "languageOf": string;
    "inScript": string;
    "select": string;
    "cancel": string;
};


export interface ILocalizedStrings {
	loaded: boolean;
	lang: string;
	languagePicker: ILanguagePickerStrings;
	[key: string]: any;
};
