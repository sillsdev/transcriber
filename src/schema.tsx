import { KeyMap, Operation, Schema, SchemaSettings } from '@orbit/data';
import Memory from '@orbit/memory';
import IndexedDBSource from '@orbit/indexeddb';
import Coordinator from '@orbit/coordinator';
import { isElectron } from './api-variable';
import { getFingerprint } from './utils';
import { offlineProjectCreate } from './crud';

const schemaDefinition: SchemaSettings = {
  models: {
    activitystate: {
      keys: { remoteId: {} },
      attributes: {
        state: { type: 'string' },
        sequencenum: { type: 'number' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {},
    },
    group: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        abbreviation: { type: 'string' },
        ownerId: { type: 'number' },
        allUsers: { type: 'boolean' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        owner: { type: 'hasOne', model: 'organization', inverse: 'groups' },
        projects: { type: 'hasMany', model: 'project', inverse: 'group' },
        groupMemberships: {
          type: 'hasMany',
          model: 'groupmembership',
          inverse: 'group',
        },
      },
    },
    groupmembership: {
      keys: { remoteId: {} },
      attributes: {
        font: { type: 'string' },
        fontSize: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        user: { type: 'hasOne', model: 'user', inverse: 'groupMemberships' },
        group: { type: 'hasOne', model: 'group', inverse: 'groupMemberships' },
        role: { type: 'hasOne', model: 'role' },
      },
    },
    integration: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        url: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        projectIntegrations: {
          type: 'hasMany',
          model: 'projectintegration',
          inverse: 'integration',
        },
      },
    },
    invitation: {
      keys: { remoteId: {} },
      attributes: {
        email: { type: 'string' },
        loginLink: { type: 'string' },
        invitedBy: { type: 'string' },
        strings: { type: 'string' },
        silId: { type: 'numer' },
        accepted: { type: 'boolean' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastUpdatedBy: { type: 'number' },
      },
      relationships: {
        organization: {
          type: 'hasOne',
          model: 'organization',
        },
        role: {
          type: 'hasOne',
          model: 'role',
        },
        group: {
          type: 'hasOne',
          model: 'group',
        },
        groupRole: {
          type: 'hasOne',
          model: 'role',
        },
        allUsersRole: {
          type: 'hasOne',
          model: 'role',
        },
      },
    },
    organization: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        slug: { type: 'string' },
        SilId: { type: 'number' },
        description: { type: 'string' },
        websiteUrl: { type: 'string' },
        logoUrl: { type: 'string' },
        publicByDefault: { type: 'boolean' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        owner: { type: 'hasOne', model: 'user' },
        users: { type: 'hasMany', model: 'user' },
        groups: { type: 'hasMany', model: 'group', inverse: 'owner' },
      },
    },
    organizationmembership: {
      keys: { remoteId: {} },
      attributes: {
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        user: {
          type: 'hasOne',
          model: 'user',
          inverse: 'organizationMemberships',
        },
        organization: { type: 'hasOne', model: 'organization' },
        role: { type: 'hasOne', model: 'role' },
      },
    },
    plan: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        organizedBy: { type: 'string' },
        tags: { type: 'string' },
        flat: { type: 'boolean' },
        sectionCount: { type: 'number' },
        slug: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        owner: { type: 'hasOne', model: 'user' },
        project: { type: 'hasOne', model: 'project', inverse: 'plans' },
        plantype: { type: 'hasOne', model: 'plantype', inverse: 'plans' },
        sections: { type: 'hasMany', model: 'section', inverse: 'plan' },
        mediafiles: { type: 'hasMany', model: 'mediafile', inverse: 'plan' },
      },
    },
    plantype: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        description: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        plans: { type: 'hasMany', model: 'plan', inverse: 'plantype' },
      },
    },
    project: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        slug: { type: 'string' },
        description: { type: 'string' },
        uilanguagebcp47: { type: 'string' },
        language: { type: 'string' },
        languageName: { type: 'string' },
        defaultFont: { type: 'string' },
        defaultFontSize: { type: 'string' },
        rtl: { type: 'boolean' },
        allowClaim: { type: 'boolean' },
        isPublic: { type: 'boolean' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
        dateExported: { type: 'date-time' },
        dateImported: { type: 'date-time' },
      },
      relationships: {
        projecttype: {
          type: 'hasOne',
          model: 'projecttype',
          inverse: 'projects',
        },
        owner: { type: 'hasOne', model: 'user' },
        organization: { type: 'hasOne', model: 'organization' },
        group: { type: 'hasOne', model: 'group', inverse: 'projects' },
        projectIntegrations: {
          type: 'hasMany',
          model: 'projectintegration',
          inverse: 'project',
        },
        // sections: { type: 'hasMany', model: 'section', inverse: 'project' },
        plans: { type: 'hasMany', model: 'plan', inverse: 'project' },
      },
    },
    projectintegration: {
      keys: { remoteId: {} },
      attributes: {
        settings: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        integration: {
          type: 'hasOne',
          model: 'integration',
          inverse: 'projectIntegrations',
        },
        project: {
          type: 'hasOne',
          model: 'project',
          inverse: 'projectIntegrations',
        },
      },
    },
    projecttype: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        description: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        projects: { type: 'hasMany', model: 'project', inverse: 'projecttype' },
      },
    },
    role: {
      keys: { remoteId: {} },
      attributes: {
        orgRole: { type: 'boolean' },
        groupRole: { type: 'boolean' },
        roleName: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        userRoles: { type: 'hasMany', model: 'organizationalmembership' },
      },
    },
    section: {
      keys: { remoteId: {} },
      attributes: {
        sequencenum: { type: 'number' },
        name: { type: 'string' },
        state: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        // projects: { type: 'hasMany', model: 'project', inverse: 'sections' },
        plan: { type: 'hasOne', model: 'plan', inverse: 'sections' },
        passages: {
          type: 'hasMany',
          model: 'passage',
          inverse: 'section',
        },
        editor: { type: 'hasOne', model: 'user' },
        transcriber: { type: 'hasOne', model: 'user' },
      },
    },
    passage: {
      keys: { remoteId: {} },
      attributes: {
        sequencenum: { type: 'number' },
        book: { type: 'string' },
        reference: { type: 'string' },
        state: { type: 'string' },
        hold: { type: 'boolean' },
        title: { type: 'string' },
        lastComment: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        mediafiles: { type: 'hasMany', model: 'mediafile', inverse: 'passage' },
        section: {
          type: 'hasOne',
          model: 'section',
          inverse: 'passages',
        },
      },
    },
    passagestatechange: {
      keys: { remoteId: {} },
      attributes: {
        state: { type: 'string' },
        comments: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        passage: { type: 'hasOne', model: 'passage' },
      },
    },
    mediafile: {
      keys: { remoteId: {} },
      attributes: {
        versionNumber: { type: 'number' },
        artifactType: { type: 'string' },
        eafUrl: { type: 'string' },
        audioUrl: { type: 'string' },
        duration: { type: 'number' },
        contentType: { type: 'string' },
        audioQuality: { type: 'string' },
        textQuality: { type: 'string' },
        transcription: { type: 'string' },
        originalFile: { type: 'string' },
        filesize: { type: 'number' },
        position: { type: 'number' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        plan: { type: 'hasOne', model: 'plan', inverse: 'mediafiles' },
        passage: { type: 'hasOne', model: 'passage', inverse: 'mediafiles' },
      },
    },
    user: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        givenName: { type: 'string' },
        familyName: { type: 'string' },
        email: { type: 'string' },
        phone: { type: 'string' },
        timezone: { type: 'string' },
        locale: { type: 'string' },
        isLocked: { type: 'boolean' },
        auth0Id: { type: 'string' },
        silUserid: { type: 'number' },
        identityToken: { type: 'string' },
        uilanguagebcp47: { type: 'string' },
        timercountUp: { type: 'boolean' },
        playbackSpeed: { type: 'number' },
        progressbarTypeid: { type: 'number' },
        avatarUrl: { type: 'string' },
        hotKeys: { type: 'string' },
        digestPreference: { type: 'number' },
        newsPreference: { type: 'boolean' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        organizationMemberships: {
          type: 'hasMany',
          model: 'organizationmembership',
          inverse: 'user',
        },
        groupMemberships: {
          type: 'hasMany',
          model: 'groupmembership',
          inverse: 'user',
        },
      },
    },
    currentuser: {
      keys: { remoteId: {} },
      attributes: {
        name: { type: 'string' },
        givenName: { type: 'string' },
        familyName: { type: 'string' },
        email: { type: 'string' },
        phone: { type: 'string' },
        timezone: { type: 'string' },
        locale: { type: 'string' },
        isLocked: { type: 'boolean' },
        auth0Id: { type: 'string' },
        silUserid: { type: 'number' },
        identityToken: { type: 'string' },
        uilanguagebcp47: { type: 'string' },
        timercountUp: { type: 'boolean' },
        playbackSpeed: { type: 'number' },
        progressbarTypeid: { type: 'number' },
        avatarUrl: { type: 'string' },
        hotKeys: { type: 'string' },
        dateCreated: { type: 'date-time' },
        dateUpdated: { type: 'date-time' },
        lastModifiedBy: { type: 'number' },
      },
      relationships: {
        organizationMemberships: {
          type: 'hasMany',
          model: 'organizationmembership',
        },
        groupMemberships: { type: 'hasMany', model: 'groupmembership' },
      },
    },
    orgdata: {
      keys: { remoteId: {} },
      attributes: {
        json: { type: 'string' },
        startnext: { type: 'number' },
      },
      relationships: {},
    },
    projdata: {
      keys: { remoteId: {} },
      attributes: {
        json: { type: 'string' },
        startnext: { type: 'number' },
        projectid: { type: 'number' },
      },
      relationships: {},
    },

    sectionpassage: {
      keys: { remoteId: {} },
      attributes: {
        data: { type: 'string' },
        planId: { type: 'number' },
        uuid: { type: 'string' },
      },
      relationships: {},
    },
  },
  version: 1,
};
if (process.env.REACT_APP_SCHEMAVERSION !== '1' && schemaDefinition.models) {
  schemaDefinition.models.offlineproject = {
    keys: { remoteId: {} },
    attributes: {
      computerfp: { type: 'string' },
      snapshotDate: { type: 'date-time' },
      offlineAvailable: { type: 'boolean' },
      exportedDate: { type: 'date-time' },
      dateCreated: { type: 'date-time' },
      dateUpdated: { type: 'date-time' },
      lastModifiedBy: { type: 'number' },
    },
    relationships: {
      project: {
        type: 'hasOne',
        model: 'project',
      },
    },
  };
  schemaDefinition.version = 2;
}

export const schema = new Schema(schemaDefinition);

export const keyMap = new KeyMap();

export const memory = new Memory({ schema, keyMap });
const findMissingModels = (schema: Schema, db: IDBDatabase) => {
  return Object.keys(schema.models).filter(
    (model) => !db.objectStoreNames.contains(model)
  );
};
const SaveOfflineProjectInfo = async (
  backup: IndexedDBSource,
  memory: Memory
) => {
  if (isElectron) {
    var t = await backup.pull((q) => q.findRecords('project'));
    const ops: Operation[] = [];
    var fingerprint = t[0].operations.length > 0 ? await getFingerprint() : '';
    t[0].operations.forEach((r: any) => {
      offlineProjectCreate(r.record, ops, memory, fingerprint, true);
    });
    await backup.push(ops);
    await memory.update(ops);
    console.log('done with upgrade to v2');
  }
};

export const backup = window.indexedDB
  ? new IndexedDBSource({
      schema,
      keyMap,
      name: 'backup',
      namespace: 'transcriber',
    })
  : ({} as IndexedDBSource);

backup.cache.migrateDB = function (db, event) {
  console.log('migrateDb', event);
  // Ensure that all models are registered
  findMissingModels(this.schema, db).forEach((model) => {
    console.log(
      `Registering IndexedDB model at version ${event.newVersion}: ${model}`
    );
    this.registerModel(db, model);
  });
  if (isElectron && event.newVersion === 2) {
    SaveOfflineProjectInfo(backup, memory);
  }
};

export const onlineCoordinator = new Coordinator();
onlineCoordinator.addSource(memory);
if (window.indexedDB) onlineCoordinator.addSource(backup);

export const offlineCoordinator = new Coordinator();
offlineCoordinator.addSource(memory);
if (window.indexedDB) offlineCoordinator.addSource(backup);
