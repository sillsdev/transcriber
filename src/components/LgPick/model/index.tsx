export * from "../langPicker/types";
export * from "../localization/model";
