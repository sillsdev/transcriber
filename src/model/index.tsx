export * from './organization';
export * from './organizationMembership';
export * from './group';
export * from './groupMembership';
export * from './integration';
export * from './invitation';
export * from './mediaFile';
export * from './project';
export * from './projectType';
export * from './plan';
export * from './planType';
export * from './section';
export * from './passage';
export * from './passageSection';
export * from './user';
export * from './role';
export * from './userPassage';
export * from './activityState';
export * from './state';
export * from '../store/book/types';
export * from '../store/localization/types';
export * from '../store/localization/model';
export * from '../store/orbit/types';
export * from '../store/upload/types';
export * from '../store/context/types';
export * from '../store/langPicker/types';
export * from '../store/media/types';
export * from '../store/paratext/types';
