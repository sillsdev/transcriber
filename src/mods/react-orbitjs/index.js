Object.defineProperty(exports, '__esModule', {
  value: true,
});
exports.withData = exports.DataProvider = undefined;

var _DataProvider = require('./components/DataProvider');

var _DataProvider2 = _interopRequireDefault(_DataProvider);

var _withData = require('./components/withData');

var _withData2 = _interopRequireDefault(_withData);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

exports.DataProvider = _DataProvider2.default;
exports.withData = _withData2.default;
