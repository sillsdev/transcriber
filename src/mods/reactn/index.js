var components_1 = require('./components');
var add_callback_1 = require('./add-callback');
var add_reducer_1 = require('./add-reducer');
var add_reducers_1 = require('./add-reducers');
var create_provider_1 = require('./create-provider');
var decorator_1 = require('./decorator');
var default_global_state_manager_1 = require('./default-global-state-manager');
var get_dispatch_1 = require('./get-dispatch');
var get_global_1 = require('./get-global');
var remove_callback_1 = require('./remove-callback');
var reset_global_1 = require('./reset-global');
var set_global_1 = require('./set-global');
var use_dispatch_1 = require('./use-dispatch');
var use_global_1 = require('./use-global');
var with_global_1 = require('./with-global');
var with_init_1 = require('./with-init');
var React = require('react');
module.exports = Object.assign(decorator_1.default, React, {
  addCallback: add_callback_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  addReducer: add_reducer_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  addReducers: add_reducers_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  Component: components_1.ReactNComponent,
  createProvider: create_provider_1.default,
  default: decorator_1.default,
  getDispatch: get_dispatch_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  getGlobal: get_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  PureComponent: components_1.ReactNPureComponent,
  removeCallback: remove_callback_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  resetGlobal: reset_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  setGlobal: set_global_1.default.bind(
    null,
    default_global_state_manager_1.default
  ),
  useDispatch: use_dispatch_1.default.bind(null, null),
  useGlobal: use_global_1.default.bind(null, null),
  withGlobal: with_global_1.default.bind(null, null),
  withInit: with_init_1.default,
});
