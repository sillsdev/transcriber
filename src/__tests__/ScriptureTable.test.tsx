import React from 'react';
import { Router } from 'react-router-dom';
import { setGlobal } from 'reactn';
import { DataProvider } from 'react-orbitjs';
import { Provider } from 'react-redux';
import configureStore from '../store';
import { createMuiTheme } from '@material-ui/core';
import Memory from '@orbit/memory';
import { schema, keyMap } from '../schema';
import history from '../history';
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ScriptureTable from '../components/ScriptureTable';
import { ActivityStates } from '../model';

const theme = createMuiTheme({});

const store = configureStore();

const memory = new Memory({ schema, keyMap });

const globals = {
  organization: null,
  orgRole: 'admin',
  project: null,
  projRole: 'admin',
  plan: null,
  tab: 0,
  group: '',
  user: null,
  lang: 'en',
  memory: memory,
  schema: schema,
  keyMap: keyMap,
};
setGlobal(globals);

const tree = (
  <DataProvider dataStore={memory}>
    <Provider store={store}>
      <Router history={history}>
        <ScriptureTable
          cols={{
            SectionSeq: 0,
            SectionnName: 1,
            PassageSeq: 2,
            Book: 3,
            Reference: 4,
            Title: 5,
          }}
        />
      </Router>
    </Provider>
  </DataProvider>
);

const addOneSection = async () => {
  memory.cache.reset();
  keyMap.reset();
  const plan = {
    type: 'plan',
    attributes: {
      name: 'Genesis',
    },
  } as any;
  schema.initializeRecord(plan);
  setGlobal({ ...globals, plan: plan.id });
  await memory.update(t => t.addRecord(plan));
  const section = {
    type: 'section',
    attributes: {
      sequencenum: 1,
      name: 'Creation',
    },
  } as any;
  schema.initializeRecord(section);
  await memory.update(t => t.addRecord(section));
  await memory.update(t =>
    t.replaceRelatedRecord({ type: 'section', id: section.id }, 'plan', {
      type: 'plan',
      id: plan.id,
    })
  );
  return section.id;
};

const addPassageToSection = async (sectionId: string) => {
  const passage = {
    type: 'passage',
    attributes: {
      sequencenum: 1,
      book: 'GEN',
      reference: '1:1-20',
      position: 0,
      state: ActivityStates.NoMedia,
      hold: false,
      title: 'Seven Days',
    },
  } as any;
  schema.initializeRecord(passage);
  await memory.update(t => [
    t.addRecord(passage),
    t.replaceRelatedRecord({ type: 'passage', id: passage.id }, 'section', {
      type: 'section',
      id: sectionId,
    }),
  ]);
  return passage.id;
};

afterEach(cleanup);

test('can render ScriptureTable snapshot', async () => {
  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Section$/i));
  expect(container.firstChild).toMatchSnapshot();
});

test('ScriptureTable renders on section line', async () => {
  await addOneSection();
  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  expect(getByText(/^Creation$/i)).toHaveTextContent('Creation');
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(2);
});

test('ScriptureTable renders a section with a passage', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  expect(getByText(/^Seven Days$/i)).toHaveTextContent('Seven Days');
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(3);
});

test('ScriptureTable AddPassage button adds a row', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  fireEvent.click(getByText(/Add Passage/i));
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(4);
  // sequence number column should be 2
  expect(body && body.children[3].children[3].textContent).toBe('2');
});

test('ScriptureTable AddPassage button adds first row', async () => {
  await addOneSection();

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  fireEvent.click(getByText(/Add Passage/i));
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(3);
  // sequence number column should be 1
  expect(body && body.children[2].children[3].textContent).toBe('1');
});

test('ScriptureTable AddSection button adds a row with the next section number', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  fireEvent.click(getByText(/Add Section/i));
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(4);
  // sequence number column should be 2
  expect(body && body.children[3].children[1].textContent).toBe('2');
});

test('ScriptureTable AddSection button adds second section when no passages in previous section', async () => {
  await addOneSection();

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  fireEvent.click(getByText(/Add Section/i));
  const body = container.querySelector('tbody');
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(3);
  // sequence number column should be 2
  expect(body && body.children[2].children[1].textContent).toBe('2');
});

test('ScriptureTable AddSection button adds first section', async () => {
  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Section$/i));
  const body = container.querySelector('tbody');
  let count = body && body.children.length;
  count = count ? count : 0;
  fireEvent.click(getByText(/Add Section/i));
  expect(body).not.toBeFalsy;
  expect(body && body.children.length).toBe(count + 1);
  // sequence number column should be 1
  expect(body && body.childNodes[1].childNodes[1].textContent).toBe('1');
});

test('ScriptureTable Delete action with nothing selected', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  fireEvent.click(getByText(/Action/i));
  await waitForElement(() => getByText(/Delete/i));
  fireEvent.click(getByText(/Delete/i));
  await waitForElement(() => getByText(/Please select row/i));
  expect(true);
});

test('ScriptureTable Select passage row', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, getAllByTestId, getByTestId } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  const box = getAllByTestId('check')[1];
  fireEvent.click(box.querySelector('input') as HTMLElement);
  await waitForElement(() => getByTestId('checked'));
  expect(true);
});

test('ScriptureTable Delete passage row gives confirmation', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, getAllByTestId, getByTestId } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  const box = getAllByTestId('check')[1];
  fireEvent.click(box.querySelector('input') as HTMLElement);
  await waitForElement(() => getByTestId('checked'));
  fireEvent.click(getByText(/Action/i));
  await waitForElement(() => getByText(/Delete/i));
  fireEvent.click(getByText(/Delete/i));
  await waitForElement(() => getByText(/Confirmation/i));
  expect(true);
});

test('ScriptureTable Delete passage row removes row', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, getAllByTestId, getByTestId, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  const box = getAllByTestId('check')[1];
  fireEvent.click(box.querySelector('input') as HTMLElement);
  await waitForElement(() => getByTestId('checked'));
  fireEvent.click(getByText(/Action/i));
  await waitForElement(() => getByText(/Delete/i));
  fireEvent.click(getByText(/Delete/i));
  await waitForElement(() => getByText(/Confirmation/i));
  fireEvent.click(getByText(/Yes/i));
  const body = container.querySelector('tbody');
  expect(body && body.children.length).toBe(2);
});

test('ScriptureTable Delete section row removes rows', async () => {
  const sectionId = await addOneSection();
  await addPassageToSection(sectionId);

  const { getByText, getAllByTestId, getByTestId, container } = render(tree);
  await waitForElement(() => getByText(/^Creation$/i));
  const box = getAllByTestId('check')[0];
  fireEvent.click(box.querySelector('input') as HTMLElement);
  await waitForElement(() => getAllByTestId('checked'));
  expect(getAllByTestId('checked').length).toBe(2);
  fireEvent.click(getByText(/Action/i));
  await waitForElement(() => getByText(/Delete/i));
  fireEvent.click(getByText(/Delete/i));
  await waitForElement(() => getByText(/Confirmation/i));
  fireEvent.click(getByText(/Yes/i));
  const body = container.querySelector('tbody');
  expect(body && body.children.length).toBe(1); // just the header
});
