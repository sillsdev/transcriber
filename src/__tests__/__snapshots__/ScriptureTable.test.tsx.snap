// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can render ScriptureTable snapshot 1`] = `
<div
  class="makeStyles-container-1"
>
  <div
    class="makeStyles-container-7"
  >
    <div
      class="makeStyles-paper-8"
    >
      <div
        class="makeStyles-actions-9"
      >
        <button
          aria-label="Add Section"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary makeStyles-button-12"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Section
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-13"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Add Passage"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary makeStyles-button-12"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Add Passage
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-13"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="Action"
          aria-owns="action-menu"
          class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary makeStyles-button-12"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Action
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-13"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="makeStyles-grow-11"
        >
          Â 
        </div>
        <button
          aria-label="Save"
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary makeStyles-button-12"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Save
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-icon-13"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V7l-4-4zm-5 16c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3zm3-10H5V5h10v4z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="data-grid-container"
        tabindex="0"
      >
        <table
          class="data-grid"
        >
          <tbody>
            <tr>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="0"
                row="0"
              >
                <span
                  class="value-viewer"
                />
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="1"
                row="0"
                style="width: 80px;"
              >
                <span
                  class="value-viewer"
                >
                  Section
                </span>
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="2"
                row="0"
                style="width: 280px;"
              >
                <span
                  class="value-viewer"
                >
                  Title
                </span>
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="3"
                row="0"
                style="width: 80px;"
              >
                <span
                  class="value-viewer"
                >
                  Passage
                </span>
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="4"
                row="0"
                style="width: 170px;"
              >
                <span
                  class="value-viewer"
                >
                  Book
                </span>
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="5"
                row="0"
                style="width: 180px;"
              >
                <span
                  class="value-viewer"
                >
                  Reference
                </span>
              </td>
              <td
                cell="[object Object]"
                class="cell read-only"
                col="6"
                row="0"
                style="width: 280px;"
              >
                <span
                  class="value-viewer"
                >
                  Description
                </span>
              </td>
            </tr>
          </tbody>
        </table>
      </span>
    </div>
    <div />
  </div>
</div>
`;
